core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java

```diff
@@ -431,16 +432,15 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
             for (final DiscoveryNode node : joiningNodes) {
                 if (node.equals(BECOME_MASTER_TASK) || node.equals(FINISH_ELECTION_NOT_MASTER_TASK)) {
                     // noop
-                } else if (currentNodes.nodeExists(node.getId())) {
+                } else if (currentNodes.nodeExists(node)) {
                     logger.debug("received a join request for an existing node [{}]", node);
                 } else {
-                    nodesChanged = true;
-                    nodesBuilder.put(node);
-                    for (DiscoveryNode existingNode : currentNodes) {
-                        if (node.getAddress().equals(existingNode.getAddress())) {
-                            nodesBuilder.remove(existingNode.getId());
-                            logger.warn("received join request from node [{}], but found existing node {} with same address, removing existing node", node, existingNode);
-                        }
+                    try {
+                        nodesBuilder.put(node);
+                        nodesChanged = true;
+                    } catch (IllegalArgumentException e) {
+                        results.failure(node, e);
```

Comment:
we get those for free from : https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java#L585